### PR TITLE
statically guarantee that current error codes are documented

### LINKED ISSUE #61137: Add missing long error code explanations
This is a re-opening of #32777.

## Explanations

Those error codes and their explanations are declared in files named `error_codes.rs`. In those files, you can see generally two macros: `register_diagnostics` and `register_long_diagnostics`. If an error code is in `register_diagnostics`, it means it lacks long error explanation. The goal of this issue is to add those missing explanations!

This is an easy first issue but don't hesitate to ask for help if you're stuck on something!

Here is the list of **all** error codes that don't have a long error explanation yet:

 * [x] E0183 (#89885)
 * [x] E0203 (#66880)
 * [x] E0208 (#98011)
 * [x] E0212
 * [x] E0224
 * [x] E0226
 * [x] E0227 (#93809)
 * [x] E0228
 * [x] E0279 (not emitted anymore)
 * [x] E0280 (#106521)
 * [x] E0284
 * [x] E0307 (#64110)
 * [x] E0311 (https://github.com/rust-lang/rust/pull/100747)
 * [x] E0312 (#64347)
 * [x] E0313 (#106580)
 * [x] E0316
 * [x] E0320 (https://github.com/rust-lang/rust/pull/105792)
 * [x] E0377 (https://github.com/rust-lang/rust/pull/105935)
 * [x] E0457 (#105860)
 * [x] E0460 (#105860)
 * [x] E0461 (https://github.com/rust-lang/rust/pull/106028)
 * [ ] E0462
 * [x] E0464 (#89587)
 * [x] E0465 (https://github.com/rust-lang/rust/pull/106282)
 * [x] E0472 (#105791)
 * [x] E0476 (#108444)
 * [x] E0477
 * [x] E0482 (#89710)
 * [x] E0490 (https://github.com/rust-lang/rust/pull/106714)
 * [x] E0493 (#64377)
 * [x] E0495 (https://github.com/rust-lang/rust/pull/64404)
 * [x] E0498 (#87521)
 * [x] E0514 (https://github.com/rust-lang/rust/pull/106236)
 * [x] E0519 (https://github.com/rust-lang/rust/pull/106236)
 * [x] E0521
 * [ ] E0523
 * [x] E0524 (#64428)
 * [x] E0531 (#64455)
 * [x] E0533 (https://github.com/rust-lang/rust/pull/64735)
 * [x] E0539 (error than can only appear inside libstd code)
 * [x] E0542 (error than can only appear inside libstd code)
 * [x] E0543 (error than can only appear inside libstd code)
 * [x] E0544 (error than can only appear inside libstd code)
 * [x] E0545 (error than can only appear inside libstd code)
 * [x] E0546 (error than can only appear inside libstd code)
 * [x] E0547 (error than can only appear inside libstd code)
 * [x] E0549 (error than can only appear inside libstd code)
 * [x] E0550 (https://github.com/rust-lang/rust/pull/64838)
 * [x] E0551 (https://github.com/rust-lang/rust/pull/64918)
 * [x] E0556 (https://github.com/rust-lang/rust/pull/65055)
 * [x] E0561 (#65098)
 * [x] E0564 (#65135) not emitted anymore
 * [x] E0566 (#65164)
 * [x] E0567 (https://github.com/rust-lang/rust/pull/65179)
 * [x] E0568 (#65205)
 * [x] E0573 (#65234)
 * [x] E0574 (#65308)
 * [x] E0575 (#65334)
 * [x] E0576 (#65390)
 * [x] E0577 (#65434)
 * [x] E0578 (#65471)
 * [x] E0584 (https://github.com/rust-lang/rust/pull/65493)
 * [x] E0587 (#65563)
 * [x] E0588 (#65591)
 * [x] E0592 (#61822)
 * [x] E0594 (#66155)
 * [x] E0623 (#66186)
 * [x] E0625 (#87715)
 * [x] E0627
 * [x] E0628 (#69867)
 * [x] E0631 (#66979)
 * [x] E0632
 * [x] E0634 (#69998)
 * [x] E0637 (#68908)
 * [x] E0640 (#98011)
 * [x] E0641
 * [x] E0657
 * [x] E0666
 * [x] E0667 (https://github.com/rust-lang/rust/pull/95364)
 * [x] E0687
 * [x] E0688
 * [x] E0693 (#69993)
 * [x] E0696
 * [x] E0697 (#65215)
 * [x] E0703
 * [x] E0706
 * [x] E0708
 * [x] E0710
 * [x] E0711 (#98011)
 * [x] E0717 (#98011)
 * [x] E0719 (#69620)
 * [x] E0722
 * [x] E0724
 * [x] E0726
 * [x] E0727
 * [x] E0728
 * [x] E0734 (https://github.com/rust-lang/rust/pull/64763)
 * [x] E0739 (#69876)
 * [x] E0740 (#65405)
 * [x] E0772

### PR DESCRIPTION
**Pull request by Ezrashaw**
Closes #61137 (that's right!)

Pretty simple refactor (often just a change from `Result<Option<&str>>` to `Result<&str>`)

r? @GuillaumeGomez (could you specially look at 53044158eff0d64673a6100f701c57b484232aca? I believe you wrote that in the first place, just want to make sure you're happy with the change)

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I'm fine with it. Thanks a lot for working on this!

@bors r+ rollup

File: compiler/rustc_error_codes/src/error_codes.rs [hunk 1]
```diff
@@ -513,7 +513,9 @@ E0790: include_str!("./error_codes/E0790.md"),
 E0791: include_str!("./error_codes/E0791.md"),
 E0792: include_str!("./error_codes/E0792.md"),
 E0793: include_str!("./error_codes/E0793.md"),
-;
+}
+
+// Undocumented removed error codes. Note that many removed error codes are documented.
```
Comment (RalfJung) [inline compiler/rustc_error_codes/src/error_codes.rs:518]: How can a removed error code be documented? I'm trying to remove an error code in https://github.com/rust-lang/rust/pull/115277 and tidy complains that the error code exists but isn't used. The only way I can see to fix that is to remove the documentation as well.

Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes.rs:518]: https://github.com/rust-lang/rust/pull/115278 should help

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes.rs:518]: This list is used for `--explain`, so even if not emitted by rustc, it's normal to have them. So you never remove an entry unless it doesn't have an error code explanation.

**Pull request merged**